dvc/rwlock.py

```diff
@@ -0,0 +1,182 @@
+from contextlib import contextmanager
+
+from voluptuous import Schema, Invalid
+
+from .exceptions import DvcException
+from .lock import LockError
+from .utils.compat import (
+    convert_to_unicode,
+    FileNotFoundError,
+    JSONDecodeError,
+    str,
+)
+from .utils.fs import relpath
+
+INFO_SCHEMA = {"pid": int, "cmd": str}
+
+SCHEMA = Schema({"write": {str: INFO_SCHEMA}, "read": {str: [INFO_SCHEMA]}})
+
+
+class RWLockFileCorruptedError(DvcException):
+    def __init__(self, path, cause):
+        super(RWLockFileCorruptedError, self).__init__(
+            "Unable to read RWLock-file '{}'. JSON structure is "
+            "corrupted".format(relpath(path)),
+            cause=cause,
+        )
+
+
+class RWLockFileFormatError(DvcException):
+    def __init__(self, path, cause):
+        super(RWLockFileFormatError, self).__init__(
+            "RWLock-file '{}' format error.".format(relpath(path)), cause=cause
+        )
+
+
+@contextmanager
+def _edit_rwlock(lock_dir):
+    path = os.path.join(lock_dir, "rwlock")
+    try:
+        with open(path, "r") as fobj:
+            lock = json.load(fobj)
+        lock = SCHEMA(convert_to_unicode(lock))
+    except FileNotFoundError:
+        lock = SCHEMA({})
+    except JSONDecodeError as exc:
+        raise RWLockFileCorruptedError(path, cause=exc)
+    except Invalid as exc:
+        raise RWLockFileFormatError(path, cause=exc)
+    lock = defaultdict(dict, lock)
+    lock["read"] = defaultdict(list, lock["read"])
+    lock["write"] = defaultdict(dict, lock["write"])
+    yield lock
+    with open(path, "w+") as fobj:
+        json.dump(lock, fobj)
+
+
+def _info_to_str(info):
+    return "  (PID {}): {}".format(info["pid"], info["cmd"])
+
+
+def _get_urls(lock, path_info):
+    ret = set(filter(path_info.overlaps, lock.keys()))
+    ret.add(path_info.url)
+    return ret
+
+
+def _check_no_writers(lock, info, path_infos):
+    for path_info in path_infos:
+        urls = _get_urls(lock["write"], path_info)
+        writers = [lock["write"].get(url) for url in urls]
+        writers = list(filter(bool, writers))
+        if not writers:
+            continue
+
+        assert len(writers) == 1
+        writer = writers[0]
+        if writer == info:
+            continue
+
+        raise LockError(
+            "'{}' is busy, it is being written to by:\n{}".format(
+                str(path_info), _info_to_str(writer)
+            )
+        )
+
+
+def _check_no_readers(lock, path_infos):
+    from itertools import chain
+
+    for path_info in path_infos:
+        urls = _get_urls(lock["read"], path_info)
+
+        readers = list(
+            chain.from_iterable([lock["read"].get(url, []) for url in urls])
+        )
+        if not readers:
+            continue
+
+        str_readers = "\n".join(_info_to_str(r) for r in readers)
+        raise LockError(
+            "'{}' is busy, it is being read by:\n{}".format(
+                str(path_info), str_readers
```

Comment:
Do we want to show which path exactly is locked by particular concurrent proc? Because those might be different from `path_info`, this might change the lines above:
```python
from funcy import select_keys

# This call might be replaced by a dict comprehension
blocking = select_keys(path_info.overlaps, lock["read"])

# `blocking` is now a subset of `lock["read"]` dict
if blocking:
    lines = ["{} by {}".format(url, ", ".join(map(_info_to_str, procs))) 
             for url, procs in blocking.items()]
    ...
```

Reply:
Not really, it will usually be visible from the command anyways.